### ISSUE TITLE
DEADLOCK on warehouse tables

### ISSUE THREAD

**Issue by giulio1979** [hashicorp/boundary#1812]
Hi, we have noticed that after a few sessions have been established (5-6) system is getting slow and at some point boundary controller states deadlocks are happening in the database:

`2022-01-03T12:02:06.481Z [TRACE] error from database adapter: location=/go/internal/db/read_writer.go:618 error="ERROR: deadlock detected (SQLSTATE 40P01)"
{"id":"y5KYT5kPcp","source":"https://hashicorp.com/boundary/kubernetes-controller","specversion":"1.0","type":"error","data":{"error":"session.(Repository).CloseConnections: unable to update connection sc_R1WVdzlitl: db.Update: unknown, unknown: error #0: ERROR: deadlock detected (SQLSTATE 40P01)","error_fields":{"Code":0,"Msg":"unable to update connection sc_R1WVdzlitl","Op":"session.(Repository).CloseConnections","Wrapped":{"Code":0,"Msg":"","Op":"db.Update","Wrapped":{"Severity":"ERROR","Code":"40P01","Message":"deadlock detected","Detail":"Process 223820 waits for ShareLock on transaction 7667150; blocked by process 223818.\nProcess 223818 waits for ShareLock on transaction 7667172; blocked by process 223820.","Hint":"See server log for query details.","Position":0,"InternalPosition":0,"InternalQuery":"","Where":"while updating tuple (17,15) in relation \"wh_session_accumulating_fact\"\nSQL statement \"with\n    session_totals (session_id, total_connection_count, total_bytes_up, total_bytes_down) as (\n      select session_id,\n             sum(connection_count),\n             sum(bytes_up),\n             sum(bytes_down)\n        from wh_session_connection_accumulating_fact\n       where session_id = p_session_id\n       group by session_id\n    )\n    update wh_session_accumulating_fact\n       set total_connection_count = session_totals.total_connection_count,\n           total_bytes_up         = session_totals.total_bytes_up,\n           total_bytes_down       = session_totals.total_bytes_down\n      from session_totals\n     where wh_session_accumulating_fact.session_id = session_totals.session_id\n    returning wh_session_accumulating_fact.*\"\nPL/pgSQL function wh_rollup_connections(wt_public_id) line 5 at SQL statement\nSQL statement \"SELECT wh_rollup_connections(new.session_id)\"\nPL/pgSQL function wh_update_session_connection() line 14 at PERFORM\nSQL statement \"update session_connection set version = $1 where public_id = $2\"\nPL/pgSQL function update_version_column(
) line 6 at EXECUTE","SchemaName":"","TableName":"","ColumnName":"","DataTypeName":"","ConstraintName":"","File":"deadlock.c","Line":1146,"Routine":"DeadLockReport"}}},"id":"e_eM24ZYnJId","version":"v0.1","op":"session.(Repository).CloseConnections","request_info":{"id":"gtraceid_ZrESKwptbHSSHdkn0o37","method":"/controller.servers.services.v1.SessionService/CloseConnection"}},"datacontentype":"application/cloudevents","time":"2022-01-03T12:02:06.481237289Z"}
{"id":"PMAS4Br8or","source":"https://hashicorp.com/boundary/kubernetes-controller","specversion":"1.0","type":"error","data":{"error":"session.(Repository).CloseConnections: db.DoTx: session.(Repository).CloseConnections: unable to update connection sc_R1WVdzlitl: db.Update: unknown, unknown: error #0: ERROR: deadlock detected (SQLSTATE 40P01)","error_fields":{"Code":0,"Msg":"","Op":"session.(Repository).CloseConnections","Wrapped":{"Code":0,"Msg":"","Op":"db.DoTx","Wrapped":{"Code":0,"Msg":"unable to update connection sc_R1WVdzlitl","Op":"session.(Repository).CloseConnections","Wrapped":{"Code":0,"Msg":"","Op":"db.Update","Wrapped":{"Severity":"ERROR","Code":"40P01","Message":"deadlock detected","Detail":"Process 223820 waits for ShareLock on transaction 7667150; blocked by process 223818.\nProcess 223818 waits for ShareLock on transaction 7667172; blocked by process 223820.","Hint":"See server log for query details.","Position":0,"InternalPosition":0,"InternalQuery":"","Where":"while updating tuple (17,15) in relation \"wh_session_accumulating_fact\"\nSQL statement \"with\n    session_totals (session_id, total_connection_count, total_bytes_up, total_bytes_down) as (\n      select session_id,\n             sum(connection_count),\n             sum(bytes_up),\n             sum(bytes_down)\n        from wh_session_connection_accumulating_fact\n       where session_id = p_session_id\n       group by session_id\n    )\n    update wh_session_accumulating_fact\n       set total_connection_count = session_totals.total_connection_count,\n           total_bytes_up         = session_totals.total_bytes_up,\n           total_bytes_down       = session_totals.total_bytes_down\n      from session_totals\n     where wh_session_accumulating_fact.session_id = session_totals.session_id\n    returning wh_session_accumulating_fact.*\"\nPL/pgSQL function wh_rollup_connections(wt_public_id) line 5 at SQL statement\nSQL statement \"SELECT wh_rollup_connections(new.session_id)\"\nPL/pgSQL function wh_updat
e_session_connection() line 14 at PERFORM\nSQL statement \"update session_connection set version = $1 where public_id = $2\"\nPL/pgSQL function update_version_column() line 6 at EXECUTE","SchemaName":"","TableName":"","ColumnName":"","DataTypeName":"","ConstraintName":"","File":"deadlock.c","Line":1146,"Routine":"DeadLockReport"}}}}},"id":"e_eM24ZYnJId","version":"v0.1","op":"session.(Repository).CloseConnections","request_info":{"id":"gtraceid_ZrESKwptbHSSHdkn0o37","method":"/controller.servers.services.v1.SessionService/CloseConnection"}},"datacontentype":"application/cloudevents","time":"2022-01-03T12:02:06.482069651Z"}
`

**To Reproduce**
1. authenticate and connect to a TCP port of an application (in my case i used a minio instance)
2. use a client application to perform activities via the tunnel (in my case I used S3CMD in a loop, put file, delete file)
3. keep adding sessions

At some point deadlocks start to occur (in my case with postgres at 4GB RAM/4CPU cores, boundary at 8GB RAM/4 CPU cores, 7-8 concurrent sessions work fine before I see failures)

**Expected behavior**
A large number of sessions can be established or the system is becoming to slow

Postgres seems overloaded when this is happening:

![image](https://user-images.githubusercontent.com/13494979/148121540-04bd5bf7-c627-49dd-849b-7e06cd323657.png)

**Comment by irenarindos** [comment]
@giulio1979 We've has been trying to recreate this deadlock but so far have been unsuccessful. Would you be able to provide more details about your configuration? Is the Postgres instance dedicated to Boundary or is it being used by other processes (ex: is Minio logging s3 events to Postgres?)

**Comment by giulio1979** [comment]
Hi @irenarindos , the postgres is dedicated to boundary. The entire setup is running in a Kubernetes cluster.

Even in my case things are working fine with 1-4 sessions. The big issues appear when I start to stress the system (starting 10 sessions concurrently).
[boundary-m2m.zip](https://github.com/hashicorp/boundary/files/7831392/boundary-m2m.zip)

I am running a container in Kubernetes to test the system (source code attached) which is actually getting the authorization code from my OIDC provider and submitting it to boundary. This way I am scaling to 30-40 sessions easy in order to test under load and simulate 10 users doing some activities.

**Comment by irenarindos** [comment]
@giulio1979 Thanks for sending sharing your code! I’m still working on this issue and have a few more questions.

How many Boundary controllers and how many workers are you using? Could you share your Boundary config file (with any sensitive information redacted)? 

How Postgres is configured within your cluster? Can you share a Dockerfile/ the output of `SHOW ALL;`?

Could you enable logging in Postgres with these settings:
`log_min_messages = info`
`log_min_error_statement = error`
`log_error_verbosity = verbose`
`log_lock_waits = true`

And then recreate the issue and share the logs. 

Thanks so much!

**Comment by giulio1979** [comment]
Hi Irena, 

here is the postgres log, I cleaned the DB and rescaled my test sessions to 20 concurrent, few minutes later things got pretty bad:

[postgres.zip](https://github.com/hashicorp/boundary/files/7842247/postgres.zip)

few minutes later I attempted to connect another 10 sessions and reach 30:

[2.zip](https://github.com/hashicorp/boundary/files/7842311/2.zip)

I am running a single controller and a single worker at the moment.

The postgres I run is also in kubernetes, provisioned via Zalando operator so I can provide you the whole manifest:

apiVersion: acid.zalan.do/v1
kind: postgresql
metadata:
  name: boundary-pg-cluster
spec:
  allowedSourceRanges:
    - 127.0.0.1/32
  databases:
    boundary: boundary
  enableMasterLoadBalancer: false
  enableReplicaLoadBalancer: false
  numberOfInstances: 2
  patroni:
    pg_hba:
      - hostssl all all 0.0.0.0/0 md5
      - host all all 0.0.0.0/0 md5
      - host replication standby 0.0.0.0/0 md5
      - hostssl replication standby 0.0.0.0/0 md5
      - local all postgres trust
      - hostssl all +zalandos 127.0.0.1/32 pam
      - hostssl all +zalandos ::1/128 pam
      - hostssl all +zalandos all pam
  postgresql:
    parameters:
      deadlock_timeout: "1000"
      max_connections: "50"
      shared_buffers: 2GB
      effective_cache_size: 6GB
      maintenance_work_mem: 512MB
      checkpoint_completion_target: "0.9"
      wal_buffers: 16MB
      default_statistics_target: "100"
      random_page_cost: "1.1"
      effective_io_concurrency: "200"
      work_mem: 10485kB
      min_wal_size: 2GB
      max_wal_size: 8GB
      max_worker_processes: "8"
      max_parallel_workers_per_gather: "4"
      max_parallel_workers: "8"
      max_parallel_maintenance_workers: "4"
      log_min_messages: "info"
      log_min_error_statement: "error"
      log_error_verbosity: "verbose"
      log_lock_waits: "true"
    version: "12"
  resources:
    limits:
      cpu: "16"
      memory: 16Gi
    requests:
      cpu: "8"
      memory: 8Gi
  teamId: boundary
  users:
    boundary: []
  volume:
    size: 20Gi

**Comment by irenarindos** [comment]
@giulio1979 Thanks for sharing your logs and config!

To help us pinpoint the cause of the deadlock, could you enable query logging by adding 
`log_statement=“all”`
to your Postgres config, then re-create the issue, and share the logs?

**Comment by giulio1979** [comment]
here you go @irenarindos 
[all-log.zip](https://github.com/hashicorp/boundary/files/7857211/all-log.zip)

**Comment by irenarindos** [comment]
Thanks so much again for sending us logs!

We’ve still been trying to re-create the deadlock and would like to better replicate the environment you’re using. Could you send us the following so we can update our configuration:
- The contents of your Boundary config file, with redacted addresses, URLs, kms configurations, database URL, and anything else you might consider sensitive information.
- All Postgres configuration parameters, by using a `SHOW ALL;` query

**Comment by giulio1979** [comment]
Hi @irenarindos since I am testing in a kubernetes environment I built a helm chart to deploy my test environment. I give you the entire helm chart I use, it contains the whole deployment.

[boundary.zip](https://github.com/hashicorp/boundary/files/7875775/boundary.zip)
[showall.txt](https://github.com/hashicorp/boundary/files/7875784/showall.txt)

Hope this helps. 

I noticed some of the queries are very expensive and do not use any indices resulting in massive disk IO, after adding one index on the table in question I managed to establish 40 sessions before collapsing.
